### PR TITLE
[SPARK-44709][CONNECT] Run ExecuteGrpcResponseSender in reattachable execute in new thread to fix flow control

### PR DESCRIPTION
**Pull request by juliuszsompolski**
### What changes were proposed in this pull request?

If executePlan / reattachExecute handling is done directly on the GRPC thread, flow control OnReady events are getting queued until after the handler returns, so OnReadyHandler never gets notified until after the handler exits.
The correct way to use it is for the handler to delegate work to another thread and exit. See https://github.com/grpc/grpc-java/issues/7361

Tidied up and added a lot of logging and statistics to ExecuteGrpcResponseSender and ExecuteResponseObserver to be able to observe this behaviour.

Followup work in https://issues.apache.org/jira/browse/SPARK-44625 is needed for cleanup of abandoned executions that will also make sure that these threads are joined.

### Why are the changes needed?

ExecuteGrpcResponseSender gets stuck waiting on grpcCallObserverReadySignal because events from OnReadyHandler do not arrive.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added extensive debugging to ExecuteGrpcResponseSender and ExecuteResponseObserver and tested and observer the behaviour of all the threads.

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 1]
```diff
@@ -58,6 +61,39 @@ private[connect] class ExecuteGrpcResponseSender[T <: MessageLite](
...
+        } finally {
+          if (!executeHolder.reattachable) {
+            // Non reattachable executions release here immediately.
+            // (Reattachable executions close release with ReleaseExecute RPC.)
+            executeHolder.close()
+          }
+        }
+      }
+    }
+    t.setDaemon(true) // can I not set daemon and just don't join?
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:93]: If you do that and the application tries to terminate it won't... You don't have to join.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 2]
```diff
@@ -58,6 +61,39 @@ private[connect] class ExecuteGrpcResponseSender[T <: MessageLite](
...
+        val e = new Exception()
+        logError(s"ON READY\n${e.getStackTrace.mkString("\n")}")
+        grpcCallObserverReadySignal.synchronized {
+          grpcCallObserverReadySignal.notifyAll()
+        }
+      })
+    }
+
+    // We run in a separate daemon thread
+    val t = new Thread {
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:80]: Please give the thread a name that is understandable.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 3]
```diff
@@ -58,6 +61,39 @@ private[connect] class ExecuteGrpcResponseSender[T <: MessageLite](
...
 
+  def run(lastConsumedStreamIndex: Long): Unit = {
+
+    // In reattachable execution, we check if grpcCallObserver is ready for sending.
+    // See sendResponse
+    if (executeHolder.reattachable && flowControl) {
+      val grpcCallObserver = grpcObserver.asInstanceOf[ServerCallStreamObserver[T]]
+      grpcCallObserver.setOnReadyHandler(() => {
+        val e = new Exception()
+        logError(s"ON READY\n${e.getStackTrace.mkString("\n")}")
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:72]: Debug code so remove?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 4]
```diff
@@ -166,7 +192,7 @@ private[connect] class ExecuteGrpcResponseSender[T <: MessageLite](
         // There is a response available to be sent.
         val sent = sendResponse(response.get.response, deadlineTimeMillis)
         if (sent) {
-          logDebug(s"Sent response index=$nextIndex.")
+          logError(s"Sent response index=$nextIndex.")
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:195]: Debug code?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala (hunk 2 above)
Comment (hvanhovell) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:80]: We really need to document the - absolutely bonkers - reason why this is needed.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala [hunk 5]
```diff
@@ -82,7 +82,7 @@ object Connect {
         "Set to 0 for unlimited.")
       .version("3.5.0")
       .timeConf(TimeUnit.MILLISECONDS)
-      .createWithDefaultString("5m")
+      .createWithDefaultString("20s")
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:85]: Debug code?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 6]
```diff
@@ -47,6 +48,8 @@ private[connect] class ExecuteGrpcResponseSender[T <: MessageLite](
   // Signal to wake up when grpcCallObserver.isReady()
   private val grpcCallObserverReadySignal = new Object
 
+  val flowControl = true
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:51]: yes... remove, or do you intend to do something with it?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteResponseObserver.scala [hunk 7]
```diff
@@ -47,7 +47,7 @@ import org.apache.spark.sql.connect.service.ExecuteHolder
  * @see
  *   attachConsumer
  */
-private[connect] class ExecuteResponseObserver[T <: MessageLite](val executeHolder: ExecuteHolder)
+private[connect] class ExecuteResponseObserver[T <: Message](val executeHolder: ExecuteHolder)
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteResponseObserver.scala:50]: Why this change?

**Comment by hvanhovell** [comment]
@juliuszsompolski couple of the fun threading questions:
1. Is the `ExecuteGrpcResponseSender` thread safe? For example is `detach()` safe? 
2. Is it possible for events to produced in a weird order?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 8]
```diff
@@ -34,7 +35,7 @@ import org.apache.spark.sql.connect.service.ExecuteHolder
  * @param grpcObserver
  *   the GRPC request StreamObserver
  */
-private[connect] class ExecuteGrpcResponseSender[T <: MessageLite](
+private[connect] class ExecuteGrpcResponseSender[T <: Message](
```
Comment (grundprinzip) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:38]: What's the reason for this particular change?

Comment (juliuszsompolski) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:38]: `ProtoUtils.abbreviate` wants a `Message`

**Comment by juliuszsompolski** [comment]
@hvanhovell 
> Is the ExecuteGrpcResponseSender thread safe? For example is detach() safe?

detach() is synchronized on the executeObserver it's attached to, like the other places in the sender... Modified the code a bit to make it more clear.

> Is it possible for events to produced in a weird order?

No. Execution pushes responses to executeObserver in order, there is a bunch of assertions that assure order of events.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteResponseObserver.scala [hunk 9]
```diff
@@ -85,12 +85,18 @@ private[connect] class ExecuteResponseObserver[T <: MessageLite](val executeHold
...
+  private var autoRemovedSize = CachedSize()
+  private var totalSize = CachedSize()
+
   /**
    * Total size of response to be held buffered after giving out with getResponse. 0 for none, any
    * value greater than 0 will buffer the response from getResponse.
    */
   private val retryBufferSize = if (executeHolder.reattachable) {
-    SparkEnv.get.conf.get(CONNECT_EXECUTE_REATTACHABLE_SENDER_MAX_STREAM_SIZE).toLong
+    SparkEnv.get.conf.get(CONNECT_EXECUTE_REATTACHABLE_OBSERVER_RETRY_BUFFER_SIZE).toLong
```
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteResponseObserver.scala:99]: this was a bug of using wrong config...

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala [hunk 10]
```diff
@@ -82,7 +93,7 @@ object Connect {
         "Set to 0 for unlimited.")
       .version("3.5.0")
       .timeConf(TimeUnit.MILLISECONDS)
-      .createWithDefaultString("5m")
+      .createWithDefaultString("2m")
```
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:96]: @hvanhovell I think 2 minutes of stream time is a better default.

**Comment by juliuszsompolski** [comment]
@hvanhovell

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging to master/3.5
